File: cpp/source/state/pilot/connected/Spectating.cpp
#include "state/pilot/connected/Spectating.hpp"

#include "event/input/connection/Received.hpp"
#include "event/input/user/OpenChannels.hpp"
#include "event/input/user/CloseChannels.hpp"

#include "event/output/ChannelsClosing.hpp"

#include "state/pilot/connected/RequestingControl.hpp"
#include "state/pilot/connected/ValidatingChannels.hpp"
#include "state/pilot/connected/Recovering.hpp"
#include "state/pilot/connected/Released.hpp"
#include "state/pilot/connected/Releasing.hpp"

#include "backend/ClientBackend.hpp"

using namespace fm;
using namespace fm::state;
using namespace fm::state::pilot;
using namespace fm::state::pilot::connected;

using namespace com::fleetmgr::interfaces;
using namespace com::fleetmgr::interfaces::facade::control;

using event::input::user::UserEvent;
using event::input::user::OpenChannels;
using event::input::user::CloseChannels;
using event::input::connection::ConnectionEvent;
using event::input::connection::Received;

using event::output::FacadeEvent;
using event::output::ChannelsClosing;

Spectating::Spectating(IState& state) :
    IState(state)
{
}

std::unique_ptr<IState> Spectating::start()
{
    return nullptr;
}

std::string Spectating::toString() const
{
    return "Spectating";
}

std::unique_ptr<IState> Spectating::handleUserEvent(const UserEvent& event)
{
    switch (event.getType())
    {
    case UserEvent::OPEN_CHANNELS:
    {
        const OpenChannels& openChannels = reinterpret_cast<const OpenChannels&>(event);
        ClientMessage message;
        message.set_command(Command::ADD_CHANNELS);
        for (const ChannelRequest& c : *openChannels.getChannels())
        {
            message.mutable_channelsrequest()->add_channels()->CopyFrom(c);
        }
        send(message);
        return nullptr;
    }

    case UserEvent::CLOSE_CHANNELS:
    {
        const CloseChannels& closeChannels = reinterpret_cast<const CloseChannels&>(event);
        backend.getChannelsHandler().closeChannels(closeChannels.getChannels());
        ClientMessage message;
        message.set_command(Command::REMOVE_CHANNELS);
        for (long c : closeChannels.getChannels())
        {
            message.mutable_channelsindication()->add_ids(c);
        }
        send(message);
        return nullptr;
    }

    case UserEvent::REQUEST_CONTROL:
        return std::make_unique<RequestingControl>(*this);

    case UserEvent::RELEASE:
        return std::make_unique<Releasing>(*this);

    default:
        return defaultEventHandle(event.toString());
    }
}

std::unique_ptr<IState> Spectating::handleConnectionEvent(const ConnectionEvent& event)
{
    switch (event.getType())
    {
    case ConnectionEvent::RECEIVED:
        return handleMessage(reinterpret_cast<const Received&>(event).getMessage());

    case ConnectionEvent::LOST:
        return std::make_unique<Recovering>(*this);

    default:
        return defaultEventHandle(event.toString());
    }
}

std::unique_ptr<IState> Spectating::handleMessage(const ControlMessage& message)
{
    switch (message.command())
    {
    case Command::ADD_CHANNELS:
        if (message.response() == Response::ACCEPTED)
        {
            std::shared_ptr<std::vector<ChannelResponse>> toOpen =
                    std::make_shared<std::vector<ChannelResponse>>();
            toOpen->reserve(message.channelsresponse().channels_size());
            for (int i = 0; i < message.channelsresponse().channels_size(); ++i)
            {
                toOpen->push_back(message.channelsresponse().channels(i));
            }
            return std::make_unique<ValidatingChannels>(*this, Role::LEADER, toOpen);
        }
        else
        {
            return defaultMessageHandle(message);
        }


    case Command::REMOVE_CHANNELS:
        if (message.response() == Response::ACCEPTED)
        {
            listener.onEvent(std::make_shared<FacadeEvent>(FacadeEvent::CHANNELS_CLOSED));
            return nullptr;
        }
        else
        {
            return defaultMessageHandle(message);
        }

    case Command::LEADER_UNREACHABLE:
        listener.onEvent(std::make_shared<FacadeEvent>(FacadeEvent::LEADER_UNREACHABLE));
        return nullptr;

    case Command::OPERATION_ENDED:
    {
        std::shared_ptr<std::vector<traffic::IChannel*>> channels = backend.getChannelsHandler().getChannels();
        listener.onEvent(std::make_shared<ChannelsClosing>(channels));
        backend.getChannelsHandler().closeAllChannels();
        ClientMessage message;
        message.set_command(Command::OPERATION_ENDED);
        message.set_response(Response::ACCEPTED);
        return std::make_unique<Released>(*this);
    }

    default:
        return defaultMessageHandle(message);
    }
}

File: java/src/main/java/com/fleetmgr/sdk/client/state/pilot/connected/Spectating.java
package com.fleetmgr.sdk.client.state.pilot.connected;

import com.fleetmgr.interfaces.AddChannelsRequest;
import com.fleetmgr.interfaces.Role;
import com.fleetmgr.interfaces.facade.control.ClientMessage;
import com.fleetmgr.interfaces.facade.control.Command;
import com.fleetmgr.interfaces.facade.control.ControlMessage;
import com.fleetmgr.interfaces.facade.control.Response;
import com.fleetmgr.sdk.client.event.input.connection.ConnectionEvent;
import com.fleetmgr.sdk.client.event.input.connection.Received;
import com.fleetmgr.sdk.client.event.input.user.CloseChannels;
import com.fleetmgr.sdk.client.event.input.user.OpenChannels;
import com.fleetmgr.sdk.client.event.input.user.UserEvent;
import com.fleetmgr.sdk.client.event.output.facade.ChannelsClosed;
import com.fleetmgr.sdk.client.event.output.facade.FacadeEvent;
import com.fleetmgr.sdk.client.event.output.facade.OperationEnded;
import com.fleetmgr.sdk.system.machine.BaseState;

import java.util.LinkedList;
import java.util.Optional;

/**
 * Created by: Bartosz Nawrot
 * Date: 23.09.2018
 * Description:
 */
public class Spectating extends BaseState {

    Spectating(BaseState state) {
        super(state);
    }

    @Override
    public Optional<BaseState> onUserEvent(UserEvent event) {
        switch (event.getType()) {
            case OPEN_CHANNELS:
                OpenChannels openChannels = (OpenChannels) event;
                send(ClientMessage.newBuilder()
                        .setCommand(Command.ADD_CHANNELS)
                        .setRequestChannels(AddChannelsRequest.newBuilder()
                                .addAllChannelId(openChannels.getChannels())
                                .build())
                        .build());
                return Optional.empty();

            case CLOSE_CHANNELS:
                CloseChannels closeChannels = (CloseChannels) event;
                backend.closeChannels(closeChannels.getChannels());
                send(ClientMessage.newBuilder()
                        .setCommand(Command.REMOVE_CHANNELS)
                        .setRequestChannels(AddChannelsRequest.newBuilder()
                                .addAllChannelId(closeChannels.getChannels())
                                .build())
                        .build());
                return Optional.empty();

            case REQUEST_CONTROL:
                return Optional.of(new RequestingControl(this));

            case RELEASE:
                return Optional.of(new Releasing(this));

            default:
                return defaultConnectionEventHandler(event.toString());
        }
    }

    @Override
    public Optional<BaseState> onConnectionEvent(ConnectionEvent event) {
        switch (event.getType()) {
            case RECEIVED:
                return handleMessage(((Received) event).getMessage());

            default:
                return defaultConnectionEventHandler(event.toString());
        }
    }

    private Optional<BaseState> handleMessage(ControlMessage message) {
        switch (message.getCommand()) {
            case ADD_CHANNELS:
                if (message.getResponse() == Response.ACCEPTED) {
                    return Optional.of(new ValidatingChannels(this, Role.PILOT,
                            message.getRequestChannels().getChannelList()));

                } else {
                    return defaultUserEventHandler(message);
                }

            case REMOVE_CHANNELS:
                if (message.getResponse() == Response.ACCEPTED) {
                    listener.onEvent(new ChannelsClosed(null));
                    return Optional.empty();

                } else {
                    return defaultUserEventHandler(message);
                }

            case LEADER_UNREACHABLE:
                listener.onEvent(new FacadeEvent(FacadeEvent.Type.LEADER_UNREACHABLE));
                return Optional.empty();

            case OPERATION_ENDED:
                listener.onEvent(new OperationEnded(new LinkedList<>(backend.getSockets().keySet())));
                backend.closeAllChannels();
                send(ClientMessage.newBuilder()
                        .setCommand(Command.OPERATION_ENDED)
                        .setResponse(Response.ACCEPTED)
                        .build());
                return Optional.of(new Released(this));

            default:
                return defaultUserEventHandler(message);
        }
    }

    @Override
    public String toString() {
        return "Spectating";
    }
}
